Provide an attribute on <model> that toggles the portal

MKevinn opened the pull request:
#### 441ea952f1d71fe44f09d3acbfc8ca0f67fd076e
<pre>
Provide an attribute on &lt;model&gt; that toggles the portal
<a href="https://bugs.webkit.org/show_bug.cgi?id=284670">https://bugs.webkit.org/show_bug.cgi?id=284670</a>
<a href="https://rdar.apple.com/141197697">rdar://141197697</a>

Reviewed by Ada Chan and Ryosuke Niwa.

Provide an attribute on &lt;model&gt; that toggles the portal. In addition, exposes
extra properties on WKPreferences so that only web views that opt in model
process and non portaled model can have the &apos;noportal&apos; attribute functional.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::createModelPlayer):
(WebCore::HTMLModelElement::attributeChanged):
(WebCore::HTMLModelElement::portal const):
(WebCore::HTMLModelElement::updatePortal):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
(WebCore::ModelPlayer::setHasPortal):
* Source/WebCore/Modules/model-element/ModelPlayer.h:
Default behavior is keeping the portal.

* Source/WebCore/html/HTMLAttributeNames.in:
noportal is the attribute name

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.messages.in:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::createLayer):
(WebKit::ModelProcessModelPlayerProxy::computeTransform):
(WebKit::ModelProcessModelPlayerProxy::updatePortalAndClipping):
Update key paths for portal and clipping. This will be fully functional until
<a href="https://rdar.apple.com/141457267">rdar://141457267</a> (Remove clipping when &lt;model&gt; doesn&apos;t have a portal).
Currently, it&apos;s blocked on clipping not toggle-able.

(WebKit::ModelProcessModelPlayerProxy::setHasPortal):
* Source/WebKit/UIProcess/API/Cocoa/WKPreferences.mm:
(-[WKPreferences _setModelElementEnabled:]):
(-[WKPreferences _modelProcessEnabled]):
(-[WKPreferences _setModelProcessEnabled:]):
(-[WKPreferences _modelElementEnabled]):
(-[WKPreferences _setNonPortaledModelEnabled:]):
(-[WKPreferences _nonPortaledModelEnabled]):
* Source/WebKit/UIProcess/API/Cocoa/WKPreferencesPrivate.h:
Exposes extra properties on WKPreferences so that only web views that opt in
model process and non portaled model can have the &apos;noportal&apos; attribute
functional.

* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::setHasPortal):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:

Canonical link: <a href="https://commits.webkit.org/288029@main">https://commits.webkit.org/288029@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/993550527bf1c6bbfcbd9b4fefe89bb1582c58ce

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81713 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1239 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35667 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86257 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32708 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83819 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1272 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9059 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/63755 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/21480 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84783 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/924 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/74368 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/44041 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/823 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/28542 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31162 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/74693 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/29144 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87696 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/80767 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8953 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/72093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9138 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/70187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/71322 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17756 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/15409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/14324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/103178 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8904 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/14436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/25058 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8745 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12268 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10553 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.h:
```diff
@@ -222,6 +222,8 @@ class HTMLModelElement final : public HTMLElement, private CachedRawResourceClie
     void environmentMapRequestResource();
     void environmentMapResetAndReject(Exception&&);
     void environmentMapResourceFinished();
+    bool portal() const;
```
Nit: I think to keep consistent with the method name in `ModelPlayer` (and also to make it clear that it's a boolean), this should be `hasPortal()`  and `updateHasPortal()`.

achan00 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.messages.in:
```diff
@@ -38,6 +38,7 @@ messages -> ModelProcessModelPlayerProxy {
     SetPaused(bool paused) -> (bool succeeded) Async
     SetCurrentTime(Seconds currentTime) -> () Async
     SetEnvironmentMap(Ref<WebCore::SharedBuffer> data)
+    SetHasPortal(bool hasPortal)
```
This should be marked with `[EnabledBy=NonPortaledModelEnabled]`

MKevinn replied:
Using the runtime check in HTMLModelElement as @rniwa suggested 🙂

MKevinn replied:
Verified locally with `nonPortaledModelEnabled` on `WKPreferences` turned off, `noportal` attribute doesn't work; and when it's on, `noportal` takes effect.

achan00 reviewed (commented):
I'd also double-check with @marcoscaceres to see if there's anything else we need to handle for non-standard attribute.

rniwa commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -500,6 +501,8 @@ void HTMLModelElement::attributeChanged(const QualifiedName& name, const AtomStr
         updateLoop();
     else if (name == environmentmapAttr)
         updateEnvironmentMap();
+    else if (name == noportalAttr)
```
You need to check whether NonPortaledModelEnabled is enabled or not here.

MKevinn replied:
Thanks. Updated to use runtime check.

rniwa commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -693,6 +698,17 @@ void HTMLModelElement::setCurrentTime(double currentTime)
 
+bool HTMLModelElement::hasPortal() const
+{
+    return !hasAttributeWithoutSynchronization(HTMLNames::noportalAttr);
```
This should probably also check `document().settings().nonPortaledModelEnabled()`?
e.g. `return document().settings().nonPortaledModelEnabled() && !hasAttributeWithoutSynchronization(HTMLNames::noportalAttr)`.

MKevinn replied:
Modified to `!(document().settings().nonPortaledModelEnabled() && hasAttributeWithoutSynchronization(HTMLNames::noportalAttr))` since this should default to true.

rniwa commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -5415,6 +5415,21 @@ NeedsStorageAccessFromFileURLsQuirk:
     WebCore:
       default: true
 
+NonPortaledModelEnabled:
```
This flag is hard to read. It's Non-Portal'ed model.
How about `ModelWithoutPortal` or `ModelWithPortalDisabled`?

rniwa replied:
Or maybe `NoPortalModelEnabled`?

MKevinn replied:
I'm leaning towards `ModelWithoutPortalEnabled` or `NoPortalModelEnabled`. cc @achan00

achan00 replied:
I like `ModelWithoutPortalEnabled`

achan00 replied:
`ModelWithoutPortal` may be simpler (worried whether there's confusion with how `ModelWithoutPortalEnabled` is parsed: ModelWithPortal-Enabled or ModelWith-PortalEnabled).

MKevinn replied:
Per offline discussion, we ended up using `ModelNoPortalAttributeEnabled`, as the name reflects its purpose clearly.

rniwa commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -693,6 +698,17 @@ void HTMLModelElement::setCurrentTime(double currentTime)
+
+void HTMLModelElement::updateHasPortal()
+{
+    if (m_modelPlayer)
```
Please store m_modelPlayer in a local variable RefPtr:
```cpp
if (RefPtr modelPlayer = m_modelPlayer)
    modelPlayer->setHasPortal(hasPortal())
```

achan00 commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -5257,6 +5257,21 @@ ModelElementEnabled:
+ModelNoPortalAttributeEnabled:
+  type: bool
+  status: unstable
+  condition: ENABLE(MODEL_PROCESS)
```
Should we add `PLATFORM(VISION)` in the condition check?

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -693,6 +698,21 @@ void HTMLModelElement::setCurrentTime(double currentTime)
+void HTMLModelElement::updateHasPortal()
+{
+    if (RefPtr modelPlayer = m_modelPlayer)
+        modelPlayer->setHasPortal(hasPortal());
```
I think `m_modelPlayer` is already a `RefPtr`. So can just do:
```
if (m_modelPlayer)
    m_modelPlayer->setHasPortal(hasPortal());
```

MKevinn replied:
Yes!

achan00 replied:
Just saw Ryosuke's comment. I think he meant to store it in a local `RefPtr` to protect its reference. Please ignore my comment. Sorry about that!

rniwa reviewed (approved):
Can we add a test?

Commit:
Provide an attribute on <model> that toggles the portal
https://bugs.webkit.org/show_bug.cgi?id=284670
rdar://141197697

Reviewed by Ada Chan and Ryosuke Niwa.

Provide an attribute on <model> that toggles the portal. In addition, exposes
extra properties on WKPreferences so that only web views that opt in model
process and non portaled model can have the 'noportal' attribute functional.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::createModelPlayer):
(WebCore::HTMLModelElement::attributeChanged):
(WebCore::HTMLModelElement::portal const):
(WebCore::HTMLModelElement::updatePortal):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
(WebCore::ModelPlayer::setHasPortal):
* Source/WebCore/Modules/model-element/ModelPlayer.h:
Default behavior is keeping the portal.

* Source/WebCore/html/HTMLAttributeNames.in:
noportal is the attribute name

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.messages.in:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::createLayer):
(WebKit::ModelProcessModelPlayerProxy::computeTransform):
(WebKit::ModelProcessModelPlayerProxy::updatePortalAndClipping):
Update key paths for portal and clipping. This will be fully functional until
rdar://141457267 (Remove clipping when <model> doesn't have a portal).
Currently, it's blocked on clipping not toggle-able.

(WebKit::ModelProcessModelPlayerProxy::setHasPortal):
* Source/WebKit/UIProcess/API/Cocoa/WKPreferences.mm:
(-[WKPreferences _setModelElementEnabled:]):
(-[WKPreferences _modelProcessEnabled]):
(-[WKPreferences _setModelProcessEnabled:]):
(-[WKPreferences _modelElementEnabled]):
(-[WKPreferences _setNonPortaledModelEnabled:]):
(-[WKPreferences _nonPortaledModelEnabled]):
* Source/WebKit/UIProcess/API/Cocoa/WKPreferencesPrivate.h:
Exposes extra properties on WKPreferences so that only web views that opt in
model process and non portaled model can have the 'noportal' attribute
functional.

* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::setHasPortal):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:

Canonical link: https://commits.webkit.org/288029@main
